### PR TITLE
Recomb pass n

### PR DESCRIPTION
**Pull request by jeromekelleher**
Here's another pass through the recomb section @petrelharp. I've reorganised things, and tried to make myself understand the rho^2 argument. You might want to have another pass through it, or we can iterate here, it's up to you.

### CODE REVIEW

**Comment by jeromekelleher** [comment]
Ah, I realise I forgot to put in how output size scales with rho and n. However, we know this exactly from the 2016 paper, so I don't know if we want to go over it again? We also say it in the "Tree Sequences" section where we say the size is O(n + t + m) for t trees, and m sites. So I think it's OK?

File: evaluation/plot.py [hunk 1]
```diff
@@ -229,7 +230,10 @@ def annotate_rho(ax, rho, x_offset, text):
             X[:, 2] = 1
             b, _, _, _ = np.linalg.lstsq(X, df["time"][ut], rcond=None)
 
-            ax.set_xlabel("Scaled recombination rate $\\rho/4 = N_e L$")
+            # JK: I don't feel strongly about this, but maybe it's less
+            # confusing.
+            # ax.set_xlabel("Scaled recombination rate $\\rho/4 = N_e L$")
+            ax.set_xlabel("$N_e L$ (= scaled recombination rate $\\rho/4$)")
```
Comment (petrelharp) [inline evaluation/plot.py:236]: ah, that is less confusing

File: paper.tex [hunk 2]
```diff
@@ -635,12 +635,15 @@ \subsection*{Simulating mutations}
...
-split the ancestral material for some lineage, creating two independent
+works backwards in time, generating common ancestor and recombination
+events and tracking their effects on segments of ancestral
+material inherited from the
+sample~\citep{hudson1983properties,hudson1990gene,kelleher2016efficient}.
+Common ancestor events merge the ancestral material of two lineages,
+and result in coalescences in the marginal trees when ancestral
+segments overlap.
+Recombination events split the ancestral material for some lineage
+at a given breakpoint, creating two independent
```
Comment (petrelharp) [inline paper.tex:646]: much better!

File: paper.tex [hunk 3]
```diff
@@ -710,63 +712,92 @@ \subsection*{Recombination}
...
+Fig.~\ref{fig-ancestry-perf} shows the observed run time for
+simulations with a variety of population size, chromosome
+length and sample sizes, and demonstrates that Eq.~\eqref{eqn-hsw-bound}
+correctly predicts the quadratic dependence on $\rho$,
+as previously conjectured~\citep[Fig.~2]{kelleher2016efficient}.
+We also see that the dependence on $n$ is quite weak,
+since increasing sample size 100-fold
+only increases run time by a factor of 2 or so. However, the
+$\log^2{n}$ factor implied by Eq.~\eqref{eqn-hsw-bound}
+(the sum is a Harmonic number, and approximated by $\log{n}$)
```
Comment (petrelharp) [inline paper.tex:755]: harmonic, not Harmonic, I think

File: paper.tex [hunk 4]
```diff
@@ -710,63 +712,92 @@ \subsection*{Recombination}
...
+is not well supported by the data except possibly at very
+large values of $\rho$, and so it seems a tighter bound
```
Comment (petrelharp) [inline paper.tex:757]: rather than "tighter bound" maybe "a different dependency on $n$"

File: paper.tex [hunk 5]
```diff
@@ -710,63 +712,92 @@ \subsection*{Recombination}
...
-panels are different---predicting the runtimes of days-long simulations
+Simulating whole chromosomes for many species is very fast,
+with 1000 samples of chromosome 1 for
+\emph{Arabidopsis thaliana} taking less than a second, and a few
+minutes for dogs and humans. However, the dependence on $\rho$
+\emph{is} quadratic, and if $\rho$ is sufficiently large simulations
+may not be feasible. For example, for
+\emph{Drosophila melanogaster} chromosome 2L, with
+$\rho\approx 3.9\times10^6$, Fig.~\ref{fig-ancestry-perf}
+predicts that simulation would require around 177 hours. For such
```
Comment (petrelharp) [inline paper.tex:776]: I think it's useful to put in the Ne and L numbers? And, putting those in I get something different:
```
$N_e \approx 1.7\times 10^6$ \citep{li2006inferring}
and a total genetic length of $\approx 0.5$ Morgans \citep{comeron2012many},
$\rho\approx 8.5\times10^5$, Fig.~\ref{fig-ancestry-perf}
 predicts that simulation would require around XXX hours. For such
```

File: paper.tex [hunk 6]
```diff
@@ -710,63 +712,92 @@ \subsection*{Recombination}
...
+particularly for moderate sample sizes.
+In practice, to predict the running time of a given
+simulation in \msprime, we recommend that users
+measure run time in a series of simulations with short genome lengths and
+the desired sample size,
+% possibly varying other parameters to check for interactions,
+% JK: why? ^
+% I'm worried people will take this as meaning "run every possible
+% combination of parameters". I think the genome length should be basically
+% independent of everything else, right?
```
Comment (petrelharp) [inline paper.tex:790]: well, I think the question is whether with varying demography it's still quadratic in rho... but yes, thinking through the argument below, I think it should be. So, I agree.

File: paper.tex [hunk 7]
```diff
@@ -777,60 +808,51 @@ \subsection*{Recombination}
...
 at this time the rate of common ancestor events is $M (M-1) / (4 N_e)$
 and the total rate of recombination is $M \ell$,
 where $\ell$ is the mean length of genome carried by each lineage
 (including ``trapped'' non-ancestral material).
-It is known that the maximum number of lineages is reached
+Since the maximum number of lineages is reached
 very quickly~\citep{nelson2020accounting},
-and at this maximum, recombination and coalescence rates are equal.
-Therefore, many (roughly half) of these lineages have at
+many (roughly half) of these lineages have at
```
Comment (petrelharp) [inline paper.tex:824]: hm - you've removed the justification for the "roughly half" statement.

Comment (jeromekelleher) [reply paper.tex:824]: We've already said we're at the time where these are equal a couple of sentences ago, right?

File: paper.tex [hunk 8]
```diff
@@ -777,60 +808,51 @@ \subsection*{Recombination}
...
 least two distant segments of ancestral material
 separated by a long segment of genome, implying that $\ell \approx L / 4$.
-This implies that $M$ is roughly $L N_e$.
-The number of linages decreases gradually,
-on the coalescent time scale---so, over roughly $2 N_e$ generations.
-Since the total rate of events when there are the maximum number of lineages
-happens is roughly $L^2 N_e / 4$,
+Then, since the rates of recombination and common ancestor events
+are equal the time the maximum number of lineages is present,
```
Comment (petrelharp) [inline paper.tex:828]: I think this is out of order - the value of M we get from just solving the equation above using this value for \ell, not any additional information about equal rates.

File: paper.tex [hunk 9]
```diff
@@ -777,60 +808,51 @@ \subsection*{Recombination}
...
-evenly on the coalescent time scale: in other words,
-more events happen in the time period from 0 to $N_e/2$ generations
-than from $N_e/2$ to $N_e$ generations,
-but the runtime is not dominated by the more recent time period.
+evenly on the coalescent time scale.
+% JK: I just found this confusing tbh.
+% in other words,
+% more events happen in the time period from 0 to $N_e/2$ generations
+% than from $N_e/2$ to $N_e$ generations,
+% but the run time is not dominated by the more recent time period.
```
Comment (petrelharp) [inline paper.tex:844]: fair enough, delete it.

File: paper.tex [hunk 10]
```diff
@@ -710,63 +712,92 @@ \subsection*{Recombination}
...
+% independent of everything else, right?
+and then predict run time by fitting a quadratic curve to genome length
+as in Fig.~\ref{fig-ancestry-perf}.
+It is important to note that the quadratic curves in the two
+panels of Fig.~\ref{fig-ancestry-perf} are different, and
+predicting the run times of days-long simulations
 using the timing of minutes-long runs is unlikely to be very accurate.
-Further experiments show that adding a $\log^2 n$ factor
-to predict runtimes across different sample sizes does not do well
-except possibly at very large values of $\rho$.
```
Comment (petrelharp) [inline paper.tex:751]: I think we do need to say something about the log^2 n or else people will think they can use it!

**Comment by petrelharp** [comment]
Looks great. A few comments - I could do some of these edits, but others I'd be reverting yours - so, want to see what you think and deal with them?

File: paper.tex (hunk 7 above)
Comment (petrelharp) [reply paper.tex:824]: True, but otherwise the 1/2 is totally mysterious.

File: paper.tex (hunk 4 above)
Comment (jeromekelleher) [reply paper.tex:757]: done

File: paper.tex (hunk 6 above)
Comment (jeromekelleher) [reply paper.tex:790]: OK, deleted

**Comment by jeromekelleher** [comment]
I've updated to fix the simple things @petrelharp and resolved the conversations above. Would you mind following up with the other issues? Particularly with the argument, I think there's something I'm not getting, so it would help if you could have one more look through it.

**Comment by jeromekelleher** [comment]
Synced on overleaf if you want to do it over there.

File: paper.tex (hunk 5 above)
Comment (jeromekelleher) [reply paper.tex:775]: I pulled the numbers from stdpopsim, they're in the plots.py script if you want to compare with yours. Showing where rho comes from seems like a good idea all right.

**Comment by petrelharp** [comment]
I'm not sure what to do about the argument - what I was saying is that the "roughly half" number comes from saying that linages coalesce and recombine at equal rates, and so spend about half the time in a state with a lot of trapped material; so, half the lineages at the time will have a lot of trapped material. I didn't want to put in a long handwavy argument for a constant that doesn't matter anyhow, but removing the one-half makes everything less precise and more confusing.

So, I'm inclined to put that small bit back the way it was - but, what is it you're not getting?

**Comment by jeromekelleher** [comment]
> So, I'm inclined to put that small bit back the way it was - but, what is it you're not getting?

I think what you've said above clears it up for me - how about a version of that? Or put back in what was there, I don't mind.

File: paper.tex (hunk 5 above)
Comment (petrelharp) [reply paper.tex:775]: whoops - yours are good -my mistake.

**Pull request merged**